[Python] Fix various issues uncovered by #12959

Tishj opened the pull request:
This PR references #12959

Namely points 1 and 5 have not been addressed by this PR.
Both of which are not trivially fixed.
1 because of the very overloaded responsibilities of the createDataFrame method.
5 because this is a problem at the Relation level.

Fixes are as follows:
- ColumnExpression can now be created with multiple names, to qualify the column with a table.
- (pyspark) DataFrame methods have been adjusted to return a non-ambiguous Column
- DuckDBPyRelation::Join has been adjusted to turn a list of strings into a USING clause
- `desc` and `asc` have been added to `spark.sql.functions`

Relation's constructor has been adjusted to take an `alias` parameter, this is now a private member of the Relation and is `const`.
Removing the need for `GetAlias`. The method has not been touched by this PR however.

Tishj committed:
accept py::args so we can directly populate the 'column_names' vector from these

Tishj committed:
fix the columnref creation

Tishj committed:
create a USING clause if a list of strings is provided

Tishj committed:
Merge remote-tracking branch 'upstream/main' into pyexpression_create_multi_part_columnref

Tishj committed:
add test with the USING clause by providing a list of strings as argument

Tishj committed:
generate a unique alias name by default for a Relation

Tishj committed:
Merge branch 'relation_default_alias' into pyexpression_create_multi_part_columnref

Tishj committed:
fix up the other points

Mytherin reviewed (approved):
Thanks! LGTM if CI passes

Tishj commented on tools/pythonpkg/tests/fast/spark/test_spark_join.py:
```diff
@@ -328,3 +328,57 @@ def test_self_join(self, dataframe_a):
+        second_join = df.join(df2, "name")
+        third_join = first_join.join(second_join, "name")
+
+        third_join.show()
```
This is point number 5 of the list of issues
@Mytherin it complains about duplicate alias, not sure how this should be fixed
```
duckdb.duckdb.BinderException: Binder Error: Duplicate alias "unnamed_relation_4c60c14f3e43605a" in query!
```

Mytherin replied:
All aliases are available after joins normally which is why this happens, e.g.:

```sql
CREATE TABLE df(name VARCHAR);
CREATE TABLE df2(name VARCHAR);
SELECT df.name, df2.name FROM df JOIN df2 USING (name);
┌─────────┬─────────┐
│  name   │  name   │
│ varchar │ varchar │
├─────────┴─────────┤
│      0 rows       │
└───────────────────┘
```

Joins preserve all child aliases, unlike subqueries that flatten all child aliases into a single alias.

Now for the Spark API we might want to wrap joins in subqueries, which would allow us to circumvent this. The only problem is this results in odd results when dealing with duplicate names across the relations, e.g.:

```sql
D CREATE TABLE df(id INT, name VARCHAR);
D CREATE TABLE df2(id INT, name VARCHAR);
D SELECT * FROM (FROM df JOIN df2 USING (id));
┌───────┬─────────┬─────────┐
│  id   │  name   │ name_1  │
│ int32 │ varchar │ varchar │
├───────┴─────────┴─────────┤
│          0 rows           │
└───────────────────────────┘
```

As you can see the original information (whether or not `name` came from `df` or from `df2`) is lost.

I'm not sure what the best solution is here - we could perhaps also automatically modify the alias in case of a conflict as a setting, and use that in the Spark API? I'm not sure what Spark does here and how it deals with conflicts in join aliases and availability of columns afterwards.

Mytherin replied:
Let's start off by just pushing a subquery after a join to flatten the namespace similar to what Spark does. We will have different behavior than Spark when it comes to duplicate names (as we will disambiguate them) but we can leave that as-is for now

Tishj committed:
fix some theorethical issues with RemoveQualificationRecursive, table name is not guaranteed to be the first part of the column_names

Tishj committed:
deduplicate column bindings when creating the TableRef from a JoinRelation

Tishj commented:
I added a fix for a potential issue around GetTableName and removing said table name.
In a couple places we assumed the table name to be at index 0 of the column_names, I've replaced it with a dedicated method that will correctly remove all qualifiers when the table name is removed.

I've mentioned it before but GetTableName is ambiguous at best because of STRUCTs, especially the hard limit of 4 `column_names` seems problematic

Tishj committed:
remove xfail

Tishj committed:
fix up 'control reaches end of non void function' nonsense

Tishj committed:
Merge remote-tracking branch 'upstream/main' into pyexpression_create_multi_part_columnref

Tishj committed:
Merge remote-tracking branch 'upstream/main' into pyexpression_create_multi_part_columnref

Tishj commented:
@Mytherin did we want to put this behavior behind a flag that we only enable in pyspark connections? Or do we not mind disambiguating problematic duplicated bindings in Joins created through the Relational API?

Tishj commented on src/main/relation/join_relation.cpp:
```diff
@@ -34,6 +37,104 @@ unique_ptr<QueryNode> JoinRelation::GetQueryNode() {
+	}
+	// If the value is not an empty string, the binding name already has a duplicate
+	// this creates a chain of name -> name_1 -> name_1_1 -> ...
+	while (!it->second.empty()) {
```
The reason this is more difficult than `DeduplicateColumns` is because in `GetDeduplicatedName` we need to look up the `old -> new` transformation.

If we for example have `[old, old_1, old]` this will be deduplicated to (replacing old->new for illustration purposes) `[new, new_1, new_2]`
If we look up `old_1` we won't know if this has been mapped to `new_1` or `new_2`

With the current solution we'll know because `old_1` will have its own entry ....

----------------------------

As I am typing this I realize that the current solution suffers from the same problem:
```c++
	string GetDeduplicatedName(const string &name) {
		auto it = bindings.find(name);
		if (it == bindings.end()) {
			// Binding not seen, safe to assume it's unique (I assume this errors later anyways?)
			return name;
		}

		// If 'second' is not empty, this is the start of a deduplication chain that we need to follow
		while (!it->second.empty()) {
			it = bindings.find(it->second);
		}
		return it->first;
	}
```

This will always return the last link of the chain, returning `new_2` for both the second `old` and `old_1`.
Though perhaps this works out fine because of the order in which the method is called?

Tishj replied:
I'll write a test for this

Tishj committed:
deduplicate the bindings, WIP, getting BinderExceptions on the deduplicated name

Tishj committed:
this should be a reference as we will be adjusting it

Tishj committed:
use the same logic as QueryResult::DeduplicateColuns

Tishj committed:
implicit conversion changed signedness

Tishj committed:
add missing include

Tishj committed:
Merge remote-tracking branch 'upstream/main' into pyexpression_create_multi_part_columnref

Tishj commented:
I had hopes that <https://github.com/duckdb/duckdb/pull/14035> would fix this without needing to deduplicate bindings, but alas

The problem that remains is that we have this structure:
```
"join": {
	"join": {
		"df": [age, name],
		"df2": [height, name]
	},
	"join": {
		"df": [age, name],
		"df2": [height, name]
	}
}
```

When we select from the produced bindings, we have ambiguities because `df` and `df2` are propagated twice.

Tishj committed:
Merge remote-tracking branch 'upstream/main' into pyexpression_create_multi_part_columnref

Tishj committed:
making the clang tidy gods happy

Mytherin commented on src/main/relation/join_relation.cpp:
```diff
@@ -35,6 +39,121 @@ unique_ptr<QueryNode> JoinRelation::GetQueryNode() {
+	unique_ptr<TableRef> DeduplicateBindings(unique_ptr<TableRef> original) {
+		auto &ref = *original;
+		switch (ref.type) {
+		case TableReferenceType::BASE_TABLE:
```
In order to simplify this PR, can we maybe remove this and get the simpler fixes in first? Then we can revisit this in the future.

Tishj committed:
rip out the binding deduplication

Tishj committed:
Merge remote-tracking branch 'upstream/main' into pyexpression_create_multi_part_columnref

Tishj committed:
add xfail to 'test_deduplicated_bindings'

Tishj committed:
Merge remote-tracking branch 'upstream/main' into pyexpression_create_multi_part_columnref

Tishj committed:
revert columnref changes

Mytherin commented:
Thanks!
